File: NeetCode/Swift/TwoPointers/Easy/88.MergeSortedArray.swift
//You are given two integer arrays nums1 and nums2, 
//sorted in non-decreasing order, and two integers m and n, 
//representing the number of elements in nums1 and nums2 respectively.
//Merge nums1 and nums2 into a single array sorted in non-decreasing order.

func merge(_ nums1: inout [Int], _ m: Int, _ nums2: [Int], _ n: Int) {
    var last = m + n - 1
    var m = m - 1
    var n = n - 1

    while last >= 0 {
        if m >= 0 && (n < 0 || nums1[m] > nums2[n]) {
            nums1[last] = nums1[m]
            m -= 1
        } else {
            nums1[last] = nums2[n]
            n -= 1
        }

        last -= 1
    }
}

print(merge([1,2,3,0,0,0], 3, [2,5,6], 3))
print(merge([1], 1, [], 0))
print(merge([0], 0, [1], 1))

File: python/Neetcode/TwoPointers/Easy/88.MergeSortedArray.py
# You are given two integer arrays nums1 and nums2, 
# sorted in non-decreasing order, and two integers m and n, 
# representing the number of elements in nums1 and nums2 respectively.
# Merge nums1 and nums2 into a single array sorted in non-decreasing order.



def merge(nums1, m, nums2, n):

    index = len(nums1) - 1

    while m > 0 and n > 0:
        if nums1[m - 1] > nums2[n - 1]:
            nums1[index] = nums1[m - 1]
            m -= 1
        else:
            nums1[index] = nums2[n - 1]
            n -= 1
        
        index -= 1

    while n > 0:
        nums1[index] = nums2[n - 1]
        n, index = n - 1, index - 1

    return nums1

print(merge([1,2,3,0,0,0], 3, [2,5,6], 3))
print(merge([1], 1, [], 0))
print(merge([0], 0, [1], 1))